WordPress/gutenberg
# Rest API: add /revisions endpoint for global styles

Parent issue:
- https://github.com/WordPress/gutenberg/issues/49601

## What?
Adds an endpoint that returns revisions for the global styles custom post.

`/wp/v2/global-styles/revisions`

Plucked from the experimental branch:
- https://github.com/WordPress/gutenberg/pull/49912

## Why?

In order to display revisions for global styles in the UI. See explorations in https://github.com/WordPress/gutenberg/issues/49601 and https://github.com/WordPress/gutenberg/pull/46667 for context.

## How?
Registering a new `/revisions` endpoint in the `Gutenberg_REST_Global_Styles_Revisions_Controller` class.
Much inspiration is taken from `WP_REST_Revisions_Controller`.

Response example:

```js
[
    {
        "author": 1,
        "date": "2023-04-23T23:29:24",
        "date_gmt": "2023-04-23T23:29:24",
        "id": 6,
        "modified": "2023-04-24T05:15:03",
        "modified_gmt": "2023-04-24T05:15:03",
        "parent": 4,
        "slug": "wp-global-styles-twentytwentythree",
        "settings": {},
        "styles": {
            "color": {
                "background": "var:preset|color|vivid-purple"
            }
        },
        "date_display": "5 hours ago (24 Apr @ 00:58)",
        "author_display_name": "admin",
        "author_avatar_url": "http://2.gravatar.com/avatar/ea8b076b398ee48b71cfaecf898c582b?s=24&d=mm&r=g"
    }
]
```

## Explanations

#### Why not extend `WP_REST_Revisions_Controller`? 

I did try this, but `WP_REST_Revisions_Controller` has a lot of custom methods and the response is geared towards revisions from posts/pages/templates that have rendered HTML content. 

Also I wanted to reduce the schema and ensure maximum flexibility for global styles, which are its own class of citizen.

## Testing Instructions

Run the tests!!!

`npm run test:unit:php -- --filter Gutenberg_REST_Global_Styles_Revisions_Controller_Test`

To test the endpoint in the site editor, fire up this branch and head over to the site editor.

In your browser's console, switch to the network tab and run the following :

```js
await wp.apiFetch( { url: `/index.php?rest_route=%2Fwp%2Fv2%2Fglobal-styles%2F${ wp.data.select('core').__experimentalGetCurrentGlobalStylesId() }%2Frevisions&_locale=user` } );
```


<img width="1421" alt="Screenshot 2023-04-21 at 1 47 40 pm" src="https://user-images.githubusercontent.com/6458278/233536375-e8d63199-3da7-4390-a692-4aac2af7e9a2.png">

Check the network request and response.
